### ISSUE TITLE
Consider joining the StyleCopAnalyzers

### ISSUE THREAD

**Issue by sharwell** [StyleCop/StyleCop#10]
Hi @citizenmatt,

I'd like to invite you to join the [DotNetAnalyzers/StyleCopAnalyzers](https://github.com/DotNetAnalyzers/StyleCopAnalyzers) project. Over the past 9 months, more than 20 community members have joined together to build a faithful reproduction of StyleCop using the .NET Compiler Platform (Roslyn). The project implements the nearly all of the original StyleCop rules and even includes code fixes for most of them, all without requiring any special tools or Visual Studio extensions. The latest versions of ReSharper already support these code fixes as well, so those users are not left out. We have over 5000 unit tests (including coverage of C# 6 features), a [live status page](http://stylecop.pdelvo.com/), and an [active discussion room on Gitter](https://gitter.im/DotNetAnalyzers/StyleCopAnalyzers?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge).

If you don't join us, please consider at least joining our chat room to discuss ways to make sure our diagnostic IDs stay in sync over time. :smile:

**Comment by meh-uk** [comment]
From my perspective that sounds like an awesome project.

**Comment by sharwell** [comment]
@meh-uk It's been remarkably successful on basically every front. Some of the contributors are *obsessed* with ensuring the code fixes are highly accurate in the most bizarre code, while others are keeping a close watch on StyleCop "classic" to make sure we only deviate from its behavior in places where the original had problems.

**Comment by meh-uk** [comment]
Is it correct that individual rules are only configurable in Visual Studio 2015?

**Comment by sharwell** [comment]
> Is it correct that individual rules are only configurable in Visual Studio 2015?

The project is built entirely on the new compiler, so you'll need to compile with Visual Studio 2015 to use it. The warnings it produces (or errors if you customize the severity) are reported in the build itself and not as a separate pass, which means errors will prevent the compiler from creating an output assembly.

Note that you can still install the StyleCop.Analyzers NuGet package in a project where some developers use Visual Studio 2013 or earlier, but it simply won't report results during the build and you won't have access to code fixes in the IDE. That said, you really should try to update to 2015; it's a tremendous improvement all-around.

**Comment by citizenmatt** [comment]
Yep, agreed on all points so far - StyleCop users are an obsessive bunch :smile:

Very nice looking project, and love the status page! My main concern is ReSharper users, mostly those not on VS2015. But I don't like the duplication of effort, and personally, I'd rather not maintain StyleCop as I don't use it. And yet we still have plenty of users not using VS2015 that want StyleCop.

Short term, getting a new ReSharper plugin for StyleCop is easily achieved without causing any harm to any of the projects. Longer term, I don't really know what's for the best.

But you can always come under the StyleCop organisation :stuck_out_tongue:

**Comment by sharwell** [comment]
@citizenmatt I understand. Perhaps a solution is the following:

1. Update this code base to support newer versions of ReSharper, and that's all. No new rules, and no need to support C# 6 (though it would still work in Visual Studio 2015 with C# 5).

2. For users that want support for C# 6 (which implies Visual Studio 2015), direct them to the project we are working on.

This reduces your workload dramatically (by not having to consider any new style rules), eliminates the duplication of effort, and gives a strong long-term direction to automated style checking and correction.

**Comment by meh-uk** [comment]
Sounds good.

**Comment by citizenmatt** [comment]
Yeah, this sounds like a good solution to me, too. My primary focus is getting ReSharper support up and running. So, in the short term, I have no plans to release a new version of StyleCop, add new rules, or whatever. I can potentially see me back porting some of your new rules to the ReSharper plugin, but that's as far as I would go (although if anyone turns up with PRs for StyleCop itself, it'll be time for another discussion…)

And I'll add info to the readme to point to the [DotNetAnalyzers/StyleCopAnalyzers](https://github.com/DotNetAnalyzers/StyleCopAnalyzers) repo and gitter room.

**Comment by markvantilburg** [comment]
When using this with resharper 9.2 (on vs2015) there should also be some instructions how to configure it...

**Comment by citizenmatt** [comment]
Good point. A VS2015 version of the ReSharper plugin should be about setting your config correctly.

**Comment by markvantilburg** [comment]
Not sure if you misread, but out of the box there is a conflict between StyleCopAnalyzers and resharper.

See https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/1346

**Comment by citizenmatt** [comment]
No, didn't misread. I'm saying that the plugin, under VS2015 would automatically configure ReSharper to be StyleCop compliant, but leave the analysis to StyleCopAnalyzers.
